File: Data_Structures_Linked_Lists/SinglyLinkedList/src/SinglyLinkedList.cs
namespace SinglyLinkedList
{
	class SinglyLinkedList {

		private int _size;
		private Node? _head;

		public int Size {
			get { return _size; }
		}
		
		public SinglyLinkedList()
		{
			this._head = null;
			this._size = 0;
		}

		public bool PrependNode(Node node)
		{
			if (node is null) return false;
			
			else if (_head is null)
			{
				_head = node;
				_size++;
				return true;
			}

			else
			{
				Node oldHead = _head;
				node.NextLink = oldHead;
				_head = node;
				_size++;
				return true;
			}
		}

		// appendNode
		public bool appendNode(Node node)
		{
			if (node is null) return false;

			else if (_head is null)
			{
				_head = node;
				_size++;
				return true;
			}
			
			else
			{
				Node endNode = TraverseToEndOfList();
				endNode.NextLink = node;
				_size++;
				return true;
			}
		}

		// removeNodeAtPosition
		// removeHead
		// removeTail
		// insertNodeAtPosition
		// clearList
		// isEmpty
		public void DisplayList()
		{
			Node temp = _head!;
			Console.Write("[");
			int moves = 0;
			while (temp != null)
			{
				if (moves == _size - 1) 
					Console.Write(temp.Data);
				else 
					Console.Write(temp.Data + ", ");

				temp = temp.NextLink;
				moves++;
			}
			Console.WriteLine("]");
		}
		

		// helpers
		private Node TraverseToEndOfList()
		{
			Node temp = _head;
			
			while (temp?.NextLink != null)
				temp = temp.NextLink;

			return temp;
		}
		// 
	}	
}

File: data-structures/02-linked-lists/list-implementations/01-singly-linked/java/src/SinglyLinkedList/SinglyLinkedList.java
package SinglyLinkedList;

public class SinglyLinkedList {

	private int size = 0;
	private Node head;
	private Node tail;
	public SinglyLinkedList() {
		size = 0;
		head = null;
	}
	
	public boolean preprendNode(Node node) {
		if (node == null) return false;
		
		if (head == null) {
			head = node;
			tail = head;
			size++;
			return true;
		}
		
		Node oldHead = head;
		head = node;
		head.nextLink = oldHead;
		size++;
		return true;
	}
	
	public boolean appendNode(Node node) {
		if (node == null) return false;
		
		head.nextLink = node;
		tail = node;
		size++;
		return true;
	}
	 
	public Node removeNodeFromHead() {
		Node nodeToReturn = head;
		head = head.nextLink;
		return nodeToReturn;
	}
	
	public Node removeNodeFromTail() {
		Node nodeToReturn = tail;
		int tailPosition = size;
		tail = traverseToNodeBeforePosition(tailPosition);
		System.out.println("New tail: " + tail.data);
		return tail;
	}

	// insertNodeAtIndex() : Node
	
	public void displayList() {
		Node temp = head;
		int moves = 0;
		
		System.out.println("[");
		while (temp.nextLink != null) {
			if (moves == size) 
				System.out.print(temp.data);
			else 
				System.out.print(temp.data + " ");
		}
		System.out.println("]");
	}
	
	public boolean clearList() {
		size = 0;
		head = null;
		tail = null;
		return true;
	}
	
	public int getListSize() {
		return size;
	}
	
	public boolean isEmpty() {
		return size == 0;
	}
	
	// -- Helpers --
	private Node traverseToNodeBeforePosition(int index) {
		 System.out.println("inside traverseToNodeBeforePosition()");
		 Node temp = head;
		 int moves = 0;
		 int targetLocation = index - 1;
		 
		 while (temp.nextLink != null && moves < targetLocation) {
			 temp = temp.nextLink;
		 }
		 
		 System.out.println(temp.data);
		 return temp;
	 }

}
